### PR TITLE
Simplify parsers v2

### LINKED ISSUE #1804: Update required guessit version to 2.1.2
### Expected behaviour:

While building Flexget, a current version of guessit, i.e. >=2.1.0, should be used along with python-dateutil-2.5.3.

### Actual behaviour:

Flexget force-uses guessit<=2.0.4, since guessit-2.0.5 introduced a dependency on python-dateutil-2.5.2, which was lifted again in guessit 2.1.0.
According to guessit docs, 2.1.0 is fully backwards compatible to 2.0.4.

### Additional information:

- Flexget Version: 2.10.35
- Python Version: 2.7.12
- Installation method: Gentoo ebuild
- OS and version: Gentoo unstable/current
- Link to crash log:

### Resolution

If I change the requirements.txt file with the following patch:
```
--- FlexGet-2.10.35.orig/requirements.txt	2017-01-26 08:12:09.000000000 +0100
+++ FlexGet-2.10.35/requirements.txt	2017-04-23 00:54:29.465410859 +0200
@@ -11,12 +11,11 @@
 jinja2
 # There is a bug in requests 2.4.0 where it leaks urllib3 exceptions
 requests>=2.8.0, <3.0
-#Guessit requires python-dateutil<=2.5.2
 python-dateutil>=2.5.3
 jsonschema>=2.0
 path.py>=8.1.1
 pathlib>=1.0
-guessit<=2.0.4
+guessit>=2.1.2
 apscheduler>=3.2.0
 terminaltables>=3.1.0
 colorclass>=2.2.0
```

Flexget builds and runs fine.

Best,
Torsten

PS: If such dependency issues aren't fixed, Linux distributions will drop support for Flexget.
Related Gentoo bug: https://bugs.gentoo.org/show_bug.cgi?id=614758

### LINKED ISSUE #1974: Unhandled error in plugin series: can only concatenate list (not "int") to list
### Steps to reproduce:
- Entries with multiple episodes: S04E14-E20 

#### Config:
```
https://github.com/crack00r/Flexget-Config/blob/master/config.yml
```
  
#### Log:
```
2017-09-25 05:04 VERBOSE series Series_Collection_DE_Stream Found different quality for The.Fosters.S04E14-E20.German.DD+51.DL.1080p.AmazonHD.x264-TVS. Was 1080p h264 dd+5.1, overriding with$
2017-09-25 05:04 CRITICAL task Series_Collection_DE_Stream BUG: Unhandled error in plugin series: can only concatenate list (not "int") to list
Traceback (most recent call last):
File "/usr/local/lib/python2.7/dist-packages/flexget/task.py", line 486, in __run_plugin
return method(*args, **kwargs)
File "/usr/local/lib/python2.7/dist-packages/flexget/event.py", line 23, in call
return self.func(*args, **kwargs)
File "/usr/local/lib/python2.7/dist-packages/flexget/plugins/filter/series.py", line 1514, in on_task_metainfo
self.parse_series(task.entries, series_name, series_config)
File "/usr/local/lib/python2.7/dist-packages/flexget/plugins/filter/series.py", line 1640, in parse_series
populate_entry_fields(entry, parsed, config)
File "/usr/local/lib/python2.7/dist-packages/flexget/plugins/filter/series.py", line 1235, in populate_entry_fields
entry['series_id'] = parser.pack_identifier
File "/usr/local/lib/python2.7/dist-packages/flexget/plugins/parsers/parser_common.py", line 553, in pack_identifier
return 'S%02dE%02d-E%02d' % (self.season, self.episode, self.episode + self.episodes - 1)
TypeError: can only concatenate list (not "int") to list
```

### Additional information:

- FlexGet version: 2.10.94
- Python version: 2.7
- Installation method: pip, global
- Using daemon (yes/no): yes
- OS and version: ubuntu 16.04

### PR DESCRIPTION
**Pull request by cvium**
### Motivation for changes:
Trying to finish @gazpachoking parser simplification branch

### Detailed changes:
- Upgraded to guessit 2.1.4
- Cleaned up @gazpachoking's old branch
- Fixed tests
### Addressed issues:
- Fixes #1974
- Fixes #1804
- Closes #2019
- Related #1398

#### To Do:

- [x] Series
- [x] Movies
- [x] Extract some of the logic from the parsers to the parse result classes
- [ ] Extract the validation in guessit to a separate function?

### CODE REVIEW

File: flexget/tests/test_metainfo.py [hunk 1]
```diff
@@ -176,6 +176,7 @@ def test_metainfo_movie(self, execute_task):
                                quality='1080p BluRay xvid',
                                id='flexget3 2004')
 
+    @pytest.mark.skip(msg='Parsers have been simplified')
     def test_metainfo_movie_with_guessit(self, execute_task):
```
Comment (cvium) [inline flexget/tests/test_metainfo.py]: Dunno what to do about this test

**Comment by cvium** [comment]
I'm considering if the ParseResult classes should have a **kwargs where parsers can set extra fields that are specific to those parsers (eg. Guessit's `is_3d`)

File: flexget/plugins/parsers/parser_common.py [hunk 2]
```diff
@@ -113,415 +104,104 @@ def normalize_name(name):
...
-    @abstractproperty
-    def title(self):
-        raise NotImplementedError
-
-    @abstractproperty
-    def complete(self):
-        raise NotImplementedError
+        # for some fucking reason it's impossible to print self.field here, if someone figures out why please
+        # tell me!
+        valid = 'INVALID'
```
Comment (liiight) [inline flexget/plugins/parsers/parser_common.py]:
```python
valid = 'OK' if self.valid else 'INVALID'
```

File: flexget/plugins/parsers/parser_guessit.py [hunk 3]
```diff
@@ -324,9 +59,69 @@ def _guessit_options(self, options):
...
+            codec = '10bit'
+
+        audio = guessit_result.get('audio_codec', '')
+        if audio == 'DTS' and guessit_result.get('audio_profile') in ['HD', 'HDMA']:
+            audio = 'dtshd'
+        elif guessit_result.get('audio_channels') == '5.1' and audio == 'AC3' or audio == 'DolbyDigital':
+            audio = 'dd5.1'
+
+        # Make sure everything are strings (guessit will return lists when there are multiples)
+        if isinstance(resolution, list):
```
Comment (liiight) [inline flexget/plugins/parsers/parser_guessit.py]: maybe loop over these with some internal func?
```python
for component in (resolution, source, codec, audio):
    if isinstance(component, list):
        component = ' '.join(component)
```

File: flexget/plugins/parsers/parser_guessit.py [hunk 4]
```diff
@@ -363,11 +169,178 @@ def parse_series(self, data, **kwargs):
...
+            if title_start != 0:
+                try:
+                    pre_title = max((match[0].end for match in guess_result.matches.values() if
+                                     match[0].end <= title_start))
+                except ValueError:
+                    pre_title = 0
+                for char in reversed(data[pre_title:title_start]):
+                    if char.isalnum() or char.isdigit():
+                        return SeriesParseResult(data=data, valid=False)
+                    if char.isspace() or char in '._':
```
Comment (liiight) [inline flexget/plugins/parsers/parser_guessit.py]: i know there's nothing wrong in iterating over a string like that but it looks kinda weird to me

File: flexget/plugins/parsers/parser_guessit.py [hunk 5]
```diff
@@ -363,11 +169,178 @@ def parse_series(self, data, **kwargs):
...
+                    valid = False
+        special = guess_result.get('episode_details', '').lower() == 'special'
+        if 'episode' not in guess_result.values_list:
+            episodes = len(guess_result.values_list.get('part', []))
+        else:
+            episodes = len(guess_result.values_list['episode'])
+        if episodes > 3:
+            valid = False
+        identified_by = kwargs.get('identified_by', 'auto')
+        id_type, id = None, None
```
Comment (liiight) [inline flexget/plugins/parsers/parser_guessit.py]: shadows builtin `id`

File: flexget/plugins/parsers/parser_guessit.py (hunk 4 above)
Comment (cvium) [reply flexget/plugins/parsers/parser_guessit.py]: Pretty normal when parsing tokens.

**Pull request merged**